datalad_dataverse/remote.py

```diff
@@ -92,111 +136,311 @@ def api(self):
         data_api = DataAccessApi(
             base_url=self.url,
             # this relies on having established the NativeApi in prepare()
             api_token=self._token,
         )
-        dataset = self.api.get_dataset(identifier=self.doi)
-
+        recorded_id = self.annex.getstate(key)
+        if recorded_id:  # empty string otherwise
+            file_id = recorded_id
+        else:
+            available_id = [i for i, p in self.files_all.items()
+                            if p == remote_file]
+            if available_id:
+                file_id = available_id[0]
+            else:
+                # This shouldn't happen, since transfer_retrieve would only be
+                # called if checkpresent returned True before.
+                raise ValueError(f"File {remote_file} is unknown to remote")
+        response = data_api.get_datafile(file_id)
         # http error handling
-        dataset.raise_for_status()
+        response.raise_for_status()
+        with open(local_file, "wb") as f:
+            f.write(response.content)
 
-        files_list = dataset.json()['data']['latestVersion']['files']
+    def transfer_store(self, key, local_file, datafile=None):
+        # TODO: double-check dataset locks. What exactly are they and is this
+        #       useful for us here?
 
-        # find the file we want to download
-        file_id = None
-        for current_file in files_list:
-            filename = current_file['dataFile']['filename']
-            if filename == key:
-                file_id = current_file['dataFile']['id']
-                break
+        if datafile is None:
+            datafile = Datafile()
+            datafile.set({'filename': key, 'label': key})
+        datafile.set({'pid': self.doi})
 
-        # error handling if file was not found on remote
-        if file_id is None:
-            raise ValueError(f"File {key} is unknown to remote")
+        self._upload_file(datafile=datafile,
+                          key=key,
+                          local_file=local_file,
+                          remote_file=key)
 
-        response = data_api.get_datafile(file_id)
-        # http error handling
-        response.raise_for_status()
-        with open(file, "wb") as f:
-            f.write(response.content)
+    def transferexport_store(self, key, local_file, remote_file):
+        remote_file = mangle_directory_names(remote_file)
+        # TODO: Pretty sure there's similar or identical restrictions to
+        # filenames; Double-check and be clear here.
+        if re.search(pattern='[^a-z0-9_\-.\\/\ ]',
+                     string=str(remote_file.parent),
+                     flags=re.ASCII | re.IGNORECASE):
+            self.annex.error(f"Invalid character in directory name of "
+                             f"{str(remote_file)}. Valid characters are a-Z, "
+                             f"0-9, '_', '-', '.', '\\', '/' and ' ' "
+                             f"(white space).")
```

Comment:
This is not an option. If filenames are incompatible, they have to be mangled. This implementation forces users to modified/fork a dataset whenever anyone wants to make an upload to dataverse.